---
title: "fastqc/cutadapt/salmon_quants script"
author: "MARGARET|LUKE|KEBASO
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


create  a directory where your reports will be stored
```{bash}

mkdir qualityreports
```
checking quality of the raw reads using fastqc
```{bash}
fastqc -o ./qualityreports *.fastq
```

Trimming adapters from your sequences depending on the quality as assessed from above

```{bash}
mkdir cutadaptresults

for file in *_R2.fastq
do
    
    file2=${file/_R2.fastq/_R1.fastq} # replace _R1.fastq in the file name with _R2.fastq
    out1=./cutadaptresults/${file}
    out2=${out1/_R2.fastq/_R1.fastq}
    cutadapt -m 20 -u 15 -q 25 -o ${out1} -p ${out2}  ${file} ${file2}
done
```
After trimming do a quality re-check before proceeding to the next steps, 
Quality re-check

```{bash}
mkdir quality-re-check
fastqc -o ./quality-re-check ./cutadaptresults/*.fastq
```

Quantification of the reads using salmon quantification tool
```{bash}
mkdir quants
```
salmon  requires a decoy aware transcriptome index when performing quantification in mapping based mode . 
Salmon indexing requires the names of the genome targets, which is extractable by using the grep command.
```{bash}
grep "^>" <(gunzip -c Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz) | cut -d " " -f 1 > decoys.txt
sed -i.bak -e 's/>//g' decoys.txt
```
Along with the list of decoys salmon also needs the concatenated transcriptome and genome reference file for index. 
The genome targets (decoys) should come after the transcriptome targets in the reference

```{bash}
cat gencode.v36.transcripts.fa.gz Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz > congentra.fa.gz
```
BUilding the index 
```{bash}
salmon index -t congentra.fa.gz -d decoys.txt -p 12 -i salmon_index --gencode
```
--gencode flag is for removing extra metdata in the target header separated by | from the gencode reference. 
salmon loop
```{bash}

for file in *_R2.fastq
do  
    samp=${file//_R2.fastq/_R1.fastq}
    echo "Processing ${file}"
    salmon quant -i salmon_index -l A \
         -1 ${file}\
         -2 ${samp}\
         -p 2 --validateMappings -o ./quants/${file//_R2.fastq/} 
done
```
